machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// relatedBuilds contains previous iterations of failed or interrupted builds related to this one.
+	// This will keep track of failed or interrupted build names
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	MachineOSBuildPrepared    BuildProgress = "BuildPrepared"
+	MachineOSBuiling          BuildProgress = "Building"
+	MachineOSBuildFailed      BuildProgress = "BuildFailed"
+	MachineOSBuildInterrupted BuildProgress = "BuildInterrupted"
+	MachineOSBuildRestarted   BuildProgress = "BuildRestarted"
+	MachineOSBuilt            BuildProgress = "Built"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+}
+
+// MachineOSObjectReference is used to track various object types related to a build.
+// We use this instead of a full fledged object refernece because we do not need all of the fields.
+type MachineOSObjectReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// osImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	OSImageURL string `json:"osImageURL"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret string `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret string `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	ContainerFile []byte `json:"containerFile"`
```

Comment:
> quibble: I think Containerfile may be preferred instead of ContainerFile.

In this context, treat me like a 4 year old, why would it be `Containerfile`? Is this some special keyword?

Reply:
@JoelSpeed I'm just used to it being referred to in the Podman docs as `Containerfile` instead of styled as `ContainerFile`. It's mostly just for the sake of consistentcy.

Reply:
@cdoern I think string would be more ergonomic since a cluster admin could just do something like: `$ oc get machineosbuild/worker -o yaml` and be able to see the Containerfile field without having to encode / decode it.

Reply:
Would be good to see if we have other APIs that refer to either a `Dockerfile/DockerFile` or a `ContianerFile/`Containerfile` for comparison. I'm 50/50 on this one, since it does appear to be a keyword, it's not a Container file per say, but a Containerfile, so it may make sense to stick with `containerfile`